File: python3/spiral_matrix_II.py
# Space : O(n**2)
# Time  : O(n**2)

class Solution:
    def updateCoor(self, x, y, direction):
        if direction == 'x+':
            return x + 1, y
        elif direction == 'x-':
            return x - 1, y
        elif direction == 'y+':
            return x, y + 1
        else:
            return x, y - 1

    def generateMatrix(self, n: int) -> List[List[int]]:
        matrix = []
        for _ in range(n):
            matrix.append([0]*n)

        i = 1
        x, y = 0, 0
        direction = "x+"
        while i <= n*n:
            matrix[y][x] = i
            if direction == "x+":
                if x == n-1 or matrix[y][x+1] > 0:
                    direction = "y+"

            if direction == "x-":
                if x == 0 or matrix[y][x-1] > 0:
                    direction = "y-"

            if direction == "y+":
                if y == n-1 or matrix[y+1][x] > 0:
                    direction = "x-"

            if direction == "y-":
                if y == 0 or matrix[y-1][x] > 0:
                    direction = "x+"

            i += 1
            x, y = self.updateCoor(x, y, direction)
        return matrix

File: golang/spiral_matrix_II.go
package main

/*
Space	: O(n^2)
Time	: O(n^2)
*/

func generateMatrix(n int) [][]int {
	ans := [][]int{}

	// init
	for i := 0; i < n; i++ {
		line := []int{}
		for j := 0; j < n; j++ {
			line = append(line, 0)
		}
		ans = append(ans, line)
	}

	// populate
	count := 1
	i, j := 0, 0
	direction := "r"
	for count <= n*n {
		ans[i][j] = count
		count++

		if direction == "r" {
			if j+1 == n || ans[i][j+1] > 0 {
				direction = "d"
			}
		} else if direction == "d" {
			if i+1 == n || ans[i+1][j] > 0 {
				direction = "l"
			}
		} else if direction == "l" {
			if j-1 == -1 || ans[i][j-1] > 0 {
				direction = "u"
			}
		} else if direction == "u" {
			if i-1 == n || ans[i-1][j] > 0 {
				direction = "r"
			}
		}

		i, j = getNextCoord(i, j, direction)
	}

	return ans
}

func getNextCoord(i int, j int, direction string) (int, int) {
	if direction == "r" {
		return i, j + 1
	} else if direction == "d" {
		return i + 1, j
	} else if direction == "l" {
		return i, j - 1
	}
	return i - 1, j
}
